static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,354 @@
+
+```js
+fetch(url, { credentials: 'include' });
+```
+
+There's nothing unusual about the above, because the request is really similar to what an `<img>` can do already.
+
+```js
+fetch(url, {
+  method: 'POST',
+  body: formData,
+});
+```
+
+There's nothing unusual about the above, because the request is really similar to what a `<form>` can already do.
+
+```js
+fetch(url, {
+  method: 'wibbley-wobbley',
+  credentials: 'include',
+  headers: {
+    fancy: 'headers',
+    'here-we': 'go',
+  },
+});
+```
+
+Ok, that's pretty unusual.
+
+What counts as 'unusual' is pretty complicated, but at a high level, if it's the kind of request that other browser APIs don't generally make, then it's unusual. At a lower level, if the request method isn't `GET`, `HEAD`, or `POST`, or it includes headers or header values that aren't part of the [safelist](https://fetch.spec.whatwg.org/#cors-safelisted-request-header), then it counts as unusual.
+
+If you try to make an unusual request, the browser first asks the other origin if it's ok to send it. This process is called a _preflight_.
+
+## Preflight request
+
+Before making the main request, the browser makes a preflight request to the destination URL with a method of `OPTIONS`, and headers like this:
+
+```
+Access-Control-Request-Method: wibbley-wobbley
+Access-Control-Request-Headers: fancy, here-we
+```
+
+- `Access-Control-Request-Method` - The HTTP method that the main request will use. This is included even if the method isn't unusual.
+- `Access-Control-Request-Headers` - The unusual headers that the main request will use. If there are no unusual headers, this header isn't sent.
+
+The preflight request never includes credentials, even if the main request will.
+
+## Preflight response
+
+The server responds to indicate whether it's happy for the main request to go ahead, using headers like this:
+
+```
+Access-Control-Allow-Methods: wibbley-wobbley
+Access-Control-Allow-Headers: fancy, here-we
+```
+
+- `Access-Control-Allow-Methods` - The _unusual_ methods to allow. This can be a comma-separated list. If the main request is to be sent without credentials, this can be `*` to allow any method.
+- `Access-Control-Allow-Headers` - The _unusual_ headers to allow. If the main request is to be sent without credentials, this can be `*` to allow any header that isn't on the [💀 FORBIDDEN LIST 💀](https://fetch.spec.whatwg.org/#forbidden-header-name).
+
+The preflight response must also pass a regular CORS check, so it needs `Access-Control-Allow-Origin`, and also `Access-Control-Allow-Credentials: true` if the main request is to be sent with credentials.
+
+If the intended method is allowed, and all the intended headers are allowed, then the main request goes ahead.
+
+Oh, and the preflight only gives the go-ahead for the request. The eventual response must also pass a CORS check.
+
+# TODO
+
+- Outro
+- Access-Control-Max-Age on preflight responses
+- Correct Vary header usage
```

Comment:
For inspiration, I was recently working with a partner who wanted CORS preflight caching applied to an entire domain. We both found [this StackOverflow response](https://stackoverflow.com/questions/12013216/how-to-apply-cors-preflight-cache-to-an-entire-domain/12021982#12021982) very insightful.